akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ProducerControllerImpl.scala

```diff
@@ -486,29 +545,92 @@ private class ProducerControllerImpl[A: ClassTag](
       s.copy(confirmedSeqNr = newMaxConfirmedSeqNr, replyAfterStore = newReplyAfterStore, unconfirmed = newUnconfirmed)
     }
 
-    def receiveStoreMessageSentCompleted(seqNr: SeqNr, m: A, ack: Boolean) = {
-      if (seqNr != s.currentSeqNr)
-        throw new IllegalStateException(s"currentSeqNr [${s.currentSeqNr}] not matching stored seqNr [$seqNr]")
+    def receiveStoreMessageSentCompleted(seqNr: SeqNr): Behavior[InternalCommand] = {
+      if (seqNr == s.storeMessageSentInProgress) {
+        if (seqNr != s.currentSeqNr)
+          throw new IllegalStateException(s"currentSeqNr [${s.currentSeqNr}] not matching stored seqNr [$seqNr]")
+        val seqMsg = s.remainingChunks.head
+        if (seqNr != seqMsg.seqNr)
+          throw new IllegalStateException(s"seqNr [${seqMsg.seqNr}] not matching stored seqNr [$seqNr]")
+
+        s.replyAfterStore.get(seqNr).foreach { replyTo =>
+          if (traceEnabled)
+            context.log.trace("Sending confirmation reply to [{}] after storage.", seqNr)
+          replyTo ! seqNr
+        }
+        val newReplyAfterStore = s.replyAfterStore - seqNr
 
-      s.replyAfterStore.get(seqNr).foreach { replyTo =>
-        if (traceEnabled)
-          context.log.trace("Sending confirmation reply to [{}] after storage.", seqNr)
-        replyTo ! seqNr
+        onMsg(seqMsg, newReplyAfterStore, s.remainingChunks.tail)
+      } else {
+        Behaviors.same
       }
-      val newReplyAfterStore = s.replyAfterStore - seqNr
+    }
 
-      onMsg(m, newReplyAfterStore, ack)
+    def receiveStoreMessageSentFailed(f: StoreMessageSentFailed[A]): Behavior[InternalCommand] = {
+      if (f.messageSent.seqNr == s.storeMessageSentInProgress) {
+        if (f.attempt >= settings.durableQueueRetryAttempts) {
+          val errorMessage =
+            s"StoreMessageSentFailed seqNr [${f.messageSent.seqNr}] failed after [${f.attempt}] attempts, giving up."
+          context.log.error(errorMessage)
+          throw new TimeoutException(errorMessage)
+        } else {
+          context.log.warnN(
+            "StoreMessageSent seqNr [{}] failed, attempt [{}] of [{}], retrying.",
+            f.messageSent.seqNr,
+            f.attempt,
+            settings.durableQueueRetryAttempts)
+          // retry
+          if (f.messageSent.isFirstChunk) {
+            storeMessageSent(f.messageSent, attempt = f.attempt + 1)
+            Behaviors.same
+          } else {
+            // store all chunks again, because partially stored chunks are discarded by the DurableQueue
+            // when it's restarted
```

Comment:
This is probably the most difficult part of this PR. When using durable queue the chunked pieces are stored rather than the full large message. The reason for that is that an event journal can most likely not store large events. It's possible that some of the chunks are stored and then something fails. Then it must take care of that when restarted.